modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/freelist/AbstractFreeList.java

```diff
@@ -267,6 +270,136 @@ private int addRowFragment(
         }
     }
 
+    /** Page handler that is used to update the expiration time of CacheDataRow. */
+    private final PageHandler<T, Boolean> updateDataRowTtl = new UpdateRowMetaHandler();
+
+    /**
+     * Page handler that is used to update the expiration time of CacheDataRow.
+     * This page handler is used only for CacheDataTree and PageIO is an instance of DataPageIO.
+     **/
+    private final class UpdateRowMetaHandler extends PageHandler<T, Boolean> {
+        /** {@inheritDoc} */
+        @Override public Boolean run(
+            int cacheId,
+            long pageId,
+            long page,
+            long pageAddr,
+            PageIO iox,
+            Boolean walPlc,
+            T row,
+            int itemId,
+            IoStatisticsHolder statHolder
+        ) throws IgniteCheckedException {
+            DataPageIO io = (DataPageIO) iox;
+
+            Boolean updated = updateTtl(pageId, page, pageAddr, io, row, itemId, walPlc, statHolder);
+
+            evictionTracker.touchPage(pageId);
+
+            return updated;
+        }
+
+        /**
+         * @param pageId Page ID.
+         * @param page Page pointer.
+         * @param pageAddr Page address.
+         * @param io IO.
+         * @param row Row.
+         * @param itemId Item id.
+         * @return {@code true} If the expiration time was updated.
+         * @throws IgniteCheckedException If failed.
+         */
+        private Boolean updateTtl(
+            long pageId,
+            long page,
+            long pageAddr,
+            DataPageIO io,
+            T row,
+            int itemId,
+            Boolean walPlc,
+            IoStatisticsHolder statHolder
+        ) throws IgniteCheckedException {
+            // Read payload to determine that the row is fragmented.
+            DataPagePayload data = io.readPayload(pageAddr, itemId, pageMem.realPageSize(grpId));
+            if (data.nextLink() == 0) {
+                // The data is not fragmented, so it is one page update.
+                io.updateExpirationTime(pageAddr, data.offset(), (CacheDataRow) row);
+
+                return Boolean.TRUE;
+            }
+
+            // The data is fragmented, so need to update the required fragment.
+            long nextLink = 0L;
+            boolean pageUpdated = false;
+            int scannedBytes = 0;
+            int updatedBytes = 0;
+
+            // We under write lock here, so it is safe to read/write the first page.
+            boolean firstPage = true;
+
+            do {
+                final int cutItemId = firstPage ? itemId : PageIdUtils.itemId(nextLink);
+                final long curPageId = firstPage ? pageId : PageIdUtils.pageId(nextLink);
+
+                final long curPage = firstPage ? page : pageMem.acquirePage(grpId, curPageId, statHolder);
```

Comment:
Can you please clarify, why it's safe to acquire write lock for the next page?

Reply:
Clarification was added.